govcd/access_control.go

```diff
@@ -456,3 +456,144 @@ func checkSanityVdcControlAccess(vdc *Vdc) error {
+
+	if tenantContext != nil {
+		client.SetCustomHeader(getTenantContextHeader(tenantContext))
+	}
+
+	err := client.ExecuteRequestWithoutResponse(catalogUrl, http.MethodPost,
+		types.PublishCatalog, "error setting catalog publishing state: %s", publishCatalog)
+
+	if tenantContext != nil {
+		client.RemoveProvidedCustomHeaders(getTenantContextHeader(tenantContext))
+	}
+
+	return err
+}
+
+// publish publishes a catalog read-only access control to all organizations
+// This operation is usually the second step for a read-only sharing to all Orgs
+func (cat *Catalog) publish(isPublished bool) error {
+	if cat.Catalog == nil {
+		return fmt.Errorf("cannot publish catalog, Object is empty")
+	}
+
+	catalogUrl := cat.Catalog.HREF
+	if catalogUrl == "nil" || catalogUrl == "" {
+		return fmt.Errorf("cannot publish catalog, HREF is empty")
+	}
+
+	tenantContext, err := cat.getTenantContext()
+	if err != nil {
+		return fmt.Errorf("cannot publish catalog, tenant context error: %s", err)
+	}
+
+	publishParameters := types.PublishCatalogParams{
+		IsPublished: &isPublished,
+	}
+	err = publishCatalog(cat.client, catalogUrl, tenantContext, publishParameters)
+	if err != nil {
+		return err
+	}
+
+	return cat.Refresh()
+}
+
+// SetReadOnlyAccessControl will create or rescind the read-only catalog sharing to all organizations
+func (cat *Catalog) SetReadOnlyAccessControl(isPublished bool) error {
+	if cat.Catalog == nil {
+		return fmt.Errorf("cannot set access control, Object is empty")
+	}
+	err := cat.SetAccessControl(&types.ControlAccessParams{
+		IsSharedToEveryone:  false,
+		EveryoneAccessLevel: addrOf(types.ControlAccessReadOnly),
+	}, true)
+	if err != nil {
+		return fmt.Errorf("error resetting access control record for catalog %s: %s", cat.Catalog.Name, err)
+	}
+	return cat.publish(isPublished)
+}
+
+// IsSharedReadOnly returns the state of the catalog read-only sharing to all organizations
+func (cat *Catalog) IsSharedReadOnly() (bool, error) {
+	accessControl, err := cat.GetAccessControl(true)
+	if err != nil {
+		return false, err
+	}
+	if accessControl.AccessSettings != nil || accessControl.IsSharedToEveryone {
+		return false, nil
+	}
+	err = cat.Refresh()
+	if err != nil {
+		return false, err
+	}
+	return cat.Catalog.IsPublished, nil
+}
+
+// IsSharedReadOnly returns the state of the catalog read-only sharing to all organizations
+func (cat *AdminCatalog) IsSharedReadOnly() (bool, error) {
+	accessControl, err := cat.GetAccessControl(true)
+	if err != nil {
+		return false, err
+	}
+	if accessControl.AccessSettings != nil || accessControl.IsSharedToEveryone {
+		return false, nil
+	}
+	err = cat.Refresh()
+	if err != nil {
+		return false, err
+	}
+	return cat.AdminCatalog.IsPublished, nil
+}
+
+// publish publishes a catalog read-only access control to all organizations
+// This operation is usually the second step for a read-only sharing to all Orgs
+func (cat *AdminCatalog) publish(isPublished bool) error {
```

Comment:
nit, but could you move the functions so the context is a bit lighter to understand? catalog.publish adminCatalog.publish catalog.IsSharedReadOnly adminCatalog.IsSharedReadOnly etc.